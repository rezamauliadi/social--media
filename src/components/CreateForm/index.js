import React, { Component } from "react";

import { Button, Form, Message } from "semantic-ui-react";

import API from "src/api";
class CreateForm extends Component {
  state = {
    userId: 1,
    body: "",
    title: "",
    invalid: false,
    loading: false
  };

  getPayload = type => {
    const { body, title, userId } = this.state;

    if (type === "posts") {
      return { title, body, userId };
    }
    return {
      body,
      name: title,
      postId: this.props.postId,
      email: "user@email.com"
    };
  };

  submitItem = async () => {
    this.setState({ loading: true, invalid: false });

    const { body, title } = this.state;
    if (!body || !title) {
      this.setState({ loading: false, invalid: true });
      return;
    }

    const { type } = this.props;
    const payload = this.getPayload(type);
    const endpoint = type === "posts" ? API.createPost : API.createComment;
    const item = await endpoint(payload);

    this.props.onSubmit(item);
    this.setState({ loading: false, title: "", body: "" });
  };

  changeValue = (_event, { name, value }) => {
    this.setState({ [name]: value });
  };

  render() {
    const { body, title, invalid, loading } = this.state;

    return (
      <Form warning={invalid}>
        <Form.TextArea
          style={{ marginBottom: "-8px" }}
          placeholder="Write your post!"
          name="body"
          value={body}
          onChange={this.changeValue}
        />
        <Form.Input
          placeholder="Give it a title!"
          name="title"
          value={title}
          onChange={this.changeValue}
        />
        <Message warning>
          It will be great if you put some words in the post and give it a cool
          title.
        </Message>
        <Button
          content="Post It!"
          labelPosition="left"
          icon="send"
          loading={loading}
          primary
          onClick={() => this.submitItem()}
        />
      </Form>
    );
  }
}

export default CreateForm;
