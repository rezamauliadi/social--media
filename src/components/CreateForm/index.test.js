import React from "react";
import { shallow } from "enzyme";

import CreateForm from "./index.js";
import API from "src/api";

describe("Component: CreateForm", () => {
  const fnMock = jest.fn();
  const props = { type: "posts", onSubmit: fnMock };

  it("should match snapshot", () => {
    const component = shallow(<CreateForm {...props} />);
    expect(component).toMatchSnapshot();
  });

  describe("button click", () => {
    it("should call submitItem when submit button clicked", () => {
      const component = shallow(<CreateForm {...props} />);
      const instance = component.instance();

      jest.spyOn(instance, "submitItem");
      component
        .find("Button")
        .at(0)
        .simulate("click");
      expect(instance.submitItem).toHaveBeenCalled();
    });
  });

  describe("changeValue", () => {
    it("should set correct state when changeValue", () => {
      const component = shallow(<CreateForm {...props} />);
      const instance = component.instance();
      const data = {
        name: "title",
        value: "foo"
      };

      instance.changeValue({}, data);
      expect(component.state("title")).toBe("foo");
    });
  });

  describe("getPayload", () => {
    it("should return correct payload (type: posts)", () => {
      const component = shallow(<CreateForm {...props} />);
      const instance = component.instance();
      const expected = { title: "bar", body: "foo", userId: 1 };

      component.setState({ title: "bar", body: "foo" });
      const payload = instance.getPayload("posts");

      expect(payload).toEqual(expected);
    });

    it("should return correct payload (type: comments)", () => {
      const component = shallow(<CreateForm {...props} />);
      const instance = component.instance();
      const expected = {
        body: "foo",
        email: "user@email.com",
        name: "bar",
        postId: 1
      };

      component.setState({ title: "bar", body: "foo" });
      const payload = instance.getPayload("comments");

      expect(payload).toEqual(expected);
    });
  });

  describe("submitItem", () => {
    it("should do submitItem (type: posts)", done => {
      const component = shallow(<CreateForm {...props} />);
      const instance = component.instance();

      jest.spyOn(API, "createPost").mockImplementation(() => {});
      jest.spyOn(instance.props, "onSubmit");
      jest.spyOn(instance, "getPayload");
      component.setState({ body: "foo", title: "bar" });

      instance.submitItem();

      process.nextTick(() => {
        expect(instance.getPayload).toHaveBeenCalled();
        expect(instance.props.onSubmit).toHaveBeenCalled();
        done();
      });
    });

    it("should do submitItem (type: comments)", done => {
      const newProps = { type: "comments", onSubmit: fnMock };
      const component = shallow(<CreateForm {...newProps} />);
      const instance = component.instance();

      jest.spyOn(API, "createComment").mockImplementation(() => {});
      jest.spyOn(instance.props, "onSubmit");
      jest.spyOn(instance, "getPayload");
      component.setState({ body: "foo", title: "bar" });

      instance.submitItem();

      process.nextTick(() => {
        expect(instance.getPayload).toHaveBeenCalled();
        expect(instance.props.onSubmit).toHaveBeenCalled();
        done();
      });
    });
  });
});
