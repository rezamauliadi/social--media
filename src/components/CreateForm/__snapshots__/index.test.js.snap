// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CreateForm should match snapshot 1`] = `
<Form
  as="form"
  warning={false}
>
  <FormTextArea
    as={[Function]}
    control={[Function]}
    name="body"
    onChange={[Function]}
    placeholder="Write your post!"
    style={
      Object {
        "marginBottom": "-8px",
      }
    }
    value=""
  />
  <FormInput
    as={[Function]}
    control={[Function]}
    name="title"
    onChange={[Function]}
    placeholder="Give it a title!"
    value=""
  />
  <Message
    warning={true}
  >
    It will be great if you put some words in the post and give it a cool title.
  </Message>
  <Button
    as="button"
    content="Post It!"
    icon="send"
    labelPosition="left"
    loading={false}
    onClick={[Function]}
    primary={true}
  />
</Form>
`;
